src/H5TSrwlock.c

```diff
@@ -0,0 +1,717 @@
+ *
+ * Return:      Non-negative on success / Negative on failure
+ *
+ *--------------------------------------------------------------------------
+ */
+herr_t
+H5TS__rw_lock_print_stats(const char *header_str, H5TS_rw_lock_stats_t *stats)
+{
+    herr_t ret_value = SUCCEED;
+
+    FUNC_ENTER_PACKAGE_NAMECHECK_ONLY
+
+    if (H5_UNLIKELY(NULL == header_str || NULL == stats))
+        HGOTO_DONE(FAIL);
+
+    fprintf(stdout, "\n\n%s\n\n", header_str);
+    fprintf(stdout, "  read_locks_granted             = %" PRId64 "\n", stats->read_locks_granted);
+    fprintf(stdout, "  read_locks_released            = %" PRId64 "\n", stats->read_locks_released);
+    fprintf(stdout, "  real_read_locks_granted        = %" PRId64 "\n", stats->real_read_locks_granted);
+    fprintf(stdout, "  real_read_locks_released       = %" PRId64 "\n", stats->real_read_locks_released);
+    fprintf(stdout, "  max_read_locks                 = %" PRId64 "\n", stats->max_read_locks);
+    fprintf(stdout, "  max_read_lock_recursion_depth  = %" PRId64 "\n", stats->max_read_lock_recursion_depth);
+    fprintf(stdout, "  read_locks_delayed             = %" PRId64 "\n", stats->read_locks_delayed);
+    fprintf(stdout, "  write_locks_granted            = %" PRId64 "\n", stats->write_locks_granted);
+    fprintf(stdout, "  write_locks_released           = %" PRId64 "\n", stats->write_locks_released);
+    fprintf(stdout, "  real_write_locks_granted       = %" PRId64 "\n", stats->real_write_locks_granted);
+    fprintf(stdout, "  real_write_locks_released      = %" PRId64 "\n", stats->real_write_locks_released);
+    fprintf(stdout, "  max_write_locks                = %" PRId64 "\n", stats->max_write_locks);
+    fprintf(stdout, "  max_write_lock_recursion_depth = %" PRId64 "\n",
+            stats->max_write_lock_recursion_depth);
+    fprintf(stdout, "  write_locks_delayed            = %" PRId64 "\n", stats->write_locks_delayed);
+    fprintf(stdout, "  max_write_locks_pending        = %" PRId64 "\n\n", stats->max_write_locks_pending);
+
+done:
+    FUNC_LEAVE_NOAPI_NAMECHECK_ONLY(ret_value)
+} /* end H5TS__rw_lock_print_stats() */
+#endif /* H5TS_ENABLE_REC_RW_LOCK_STATS */
+
+/*--------------------------------------------------------------------------
+ * Function:    H5TS__rw_lock_init
+ *
+ * Purpose:     Initialize the supplied instance of H5TS_rw_lock_t.
+ *
+ * Return:      Non-negative on success / Negative on failure
+ *
+ *--------------------------------------------------------------------------
+ */
+herr_t
+H5TS__rw_lock_init(H5TS_rw_lock_t *rw_lock)
+{
+    herr_t ret_value = SUCCEED;
+
+    FUNC_ENTER_PACKAGE_NAMECHECK_ONLY
+
+    if (H5_UNLIKELY(NULL == rw_lock))
+        HGOTO_DONE(FAIL);
+
+#ifdef H5_HAVE_WIN_THREADS
+    /* The current H5TS_rw_lock_t implementation uses H5TS_key_create() with a
+     * key destructor callback, which is not [currently] supported by Windows.
+     */
+    HGOTO_DONE(FAIL);
```

Comment:
It's somewhat hard to find through the GitHub UI, but is `H5TS_rw_lock_t` used anywhere currently? Or rather, will this be a problem anywhere when Win32 threads are what's available?

Reply:
No, the recursive R/W lock is not used for anything yet.  I'm on the verge of deleting it, since it won't work as the API lock and I don't foresee wanting the recursion anywhere else.